File: vue-app/src/interfaces/IModelController.ts
import { AxiosStatic } from "axios";
import { ISearch } from "./ISearch";
import { IModel } from "./IModel";

export interface IModelController<dto extends IModel, search extends ISearch> {
  axiosInstance: AxiosStatic;
  ReadAll(): Promise<dto[]>;
  Read(id: number): Promise<dto>;
  Create(dto: dto): Promise<dto>;
  Update(dto: dto): Promise<dto>;
  Delete(id: number): Promise<void>;
  Search(search: search): Promise<dto[]>;
}

File: WebApi/Controllers/IModelController.cs
using Microsoft.AspNetCore.Mvc;

public interface IModelController<Dto, SearchT> where Dto : IDTO where SearchT : ISearch
{
    public Task<ActionResult<List<Dto>>> ReadAll();
    public Task<ActionResult<Dto>> Read(int id);
    public Task<ActionResult<Dto>> Create([FromBody] Dto dto);
    public Task<ActionResult<Dto>> Update([FromBody] Dto dto);
    public Task<ActionResult> Delete(int id);
    public Task<ActionResult<List<Dto>>> Search([FromBody] SearchT search);
}
